### PR TITLE
Feat/1786 precommit hooks

### LINKED ISSUE #1786: As a developer, I want linters to run automatically before git push
**Description:**
_Provide a brief background and justification for this issue_
Using git's pre-commit checks, ensure linters are auto-run before commit/push cycle.

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [ ] Allow linting errors in commits for but not pushes.
- [x] Configuration document for developers is produced
- [x] Secrets checks run before commits are allowed
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [x] Investigate git commit/push file actions
- [ ] Investigate git secrets catches sendgrid api per recent events
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_

### PR DESCRIPTION
**Pull request by andrew-jameson**
## Summary of Changes
_Provide a brief summary of changes_
Pull request closes #1786 

## How to Test
_List the steps to test the PR_
```
task gitcfg
echo "    smtp_auth_password: 409ut3445109j3g3==" >> tdrs-backend/plg/alertmanager/alertmanager.yml

git add . && git commit -m "secret failures"
#watch this catch secrets and fail the commit

echo " }@" >> tdrs-frontend/src/actions/alert.js
echo "}" >> tdrs-backend/tdpservice/users/models.py
git add . && git commit -m "." && git push
#watch push catch lint issues only on push
```

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [ ] [**_insert ACs here_**]
+ [ ] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: scripts/git-secrets-check.sh [hunk 1]
```diff
@@ -1,29 +1,53 @@
...
+    echo "git-secrets-check.sh: Scanning files staged for commit ..."
+    setopt shwordsplit
+    staged_files=$(git diff --cached --name-status | grep -vE "^R[0-9]+"| cut -f2 | xargs)
 
-echo "git-secrets-check.sh: Scanning repo ..."
-git secrets --scan -r ../
-retVal=$?
+    for filename in $staged_files; do
+        echo "git-secrets-check.sh: Scanning $filename ..."
+        git secrets --scan $filename
```
Comment (andrew-jameson) [inline scripts/git-secrets-check.sh:42]: Ensure pathing is correct since this runs in scripts directory.

File: .githooks/pre-commit [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+set -e
+
+# git-secrets needs to be installed for macos
+#git secrets --pre_commit_hook -- "$@"
+#git secrets --prepare_commit_msg_hook -- "$@"
+zsh ./scripts/git-secrets-check.sh local
+#branch=$(git symbolic-ref --short HEAD)
+#bash ./scripts/trufflehog-check.sh "$branch"
```
Comment (andrew-jameson) [inline .githooks/pre-commit]: Still a WIP.

File: .githooks/pre-commit [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+set -e
+
+# git-secrets needs to be installed for macos
+#git secrets --pre_commit_hook -- "$@"
+#git secrets --prepare_commit_msg_hook -- "$@"
+zsh ./scripts/git-secrets-check.sh local
```
Comment (jtimpe) [inline .githooks/pre-commit:4]: i get an error for this

```bash
The command git-secrets is available
git-secrets-check.sh: Scanning files staged for commit ...
git-secrets-check.sh: Scanning tdrs-backend/plg/alertmanager/alertmanager.yml ...
git: 'secrets' is not a git command. See 'git --help'.
```

Comment (andrew-jameson) [reply .githooks/pre-commit:4]: New commit should address this.

File: scripts/git-secrets-check.sh (hunk 1 above)
Comment (andrew-jameson) [reply scripts/git-secrets-check.sh:42]: All good, paths are included from `git diff` command.

**Comment by jtimpe** [comment]
the command itself works now, but i don't believe this result is correct

```bash
✓ Installed commit-msg hook to .git/hooks/commit-msg
✓ Installed pre-commit hook to .git/hooks/pre-commit
✓ Installed prepare-commit-msg hook to .git/hooks/prepare-commit-msg
git-secrets-check.sh: Scanning files staged for commit ...
git-secrets-check.sh: Scanning tdrs-backend/plg/alertmanager/alertmanager.yml ...
git-secrets-check.sh: Scanning tdrs-backend/tdpservice/users/models.py ...
git-secrets-check.sh: Scanning tdrs-frontend/src/actions/alert.js ...
git-secrets-check.sh: No issues found
[feat/1786-precommit-hooks 6256e37be] moar good code
 3 files changed, 3 insertions(+)
```

image

**Comment by andrew-jameson** [comment]
> the command itself works now, but i don't believe this result is correct
> 
> ```shell
> ✓ Installed commit-msg hook to .git/hooks/commit-msg
> ✓ Installed pre-commit hook to .git/hooks/pre-commit
> ✓ Installed prepare-commit-msg hook to .git/hooks/prepare-commit-msg
> git-secrets-check.sh: Scanning files staged for commit ...
> git-secrets-check.sh: Scanning tdrs-backend/plg/alertmanager/alertmanager.yml ...
> git-secrets-check.sh: Scanning tdrs-backend/tdpservice/users/models.py ...
> git-secrets-check.sh: Scanning tdrs-frontend/src/actions/alert.js ...
> git-secrets-check.sh: No issues found
> [feat/1786-precommit-hooks 6256e37be] moar good code
>  3 files changed, 3 insertions(+)
> ```


Wow, this was a relative path oddity! This prior logic works in CircleCI but not locally on Mac. New commit should address this, we'll see if it breaks CircleCI.

**Comment by jtimpe** [comment]
this is the error i'm seeing now (no secrets in the repo, but i added the syntax error changes to trigger the linter)

```bash
git push
error: failed to push some refs to 'github.com:raft-tech/TANF-app.git'
```

File: Taskfile.yml [hunk 4]
```diff
@@ -146,7 +151,7 @@ tasks:
     dir: tdrs-backend
     cmds:
       - task backend-up
-      - docker-compose -f docker-compose.yml exec web sh -c "flake8 . && if [ $? -eq 0 ]; then echo 'Flake8 linter found no issues'; fi"
+      - docker-compose -f docker-compose.yml exec -T web sh -c "flake8 . && if [ $? -eq 0 ]; then echo 'Flake8 linter found no issues'; fi"
```
Comment (jtimpe) [inline Taskfile.yml]:
```suggestion
      - docker compose -f docker-compose.yml exec -T web sh -c "flake8 . && if [ $? -eq 0 ]; then echo 'Flake8 linter found no issues'; fi"
```

File: Taskfile.yml [hunk 5]
```diff
@@ -234,7 +239,7 @@ tasks:
     dir: tdrs-frontend
     cmds:
       - docker-compose -f docker-compose.local.yml up -d tdp-frontend-test --quiet-pull
-      - docker-compose -f docker-compose.yml exec tdp-frontend-test sh -c "npm run lint"
+      - docker-compose -f docker-compose.yml exec -T tdp-frontend-test sh -c "npm run lint"
```
Comment (jtimpe) [inline Taskfile.yml]:
```suggestion
      - docker compose -f docker-compose.yml exec -T tdp-frontend-test sh -c "npm run lint"
```

File: Taskfile.yml [hunk 6]
```diff
@@ -234,7 +239,7 @@ tasks:
     dir: tdrs-frontend
     cmds:
       - docker-compose -f docker-compose.local.yml up -d tdp-frontend-test --quiet-pull
```
Comment (jtimpe) [inline Taskfile.yml]:
```suggestion
      - docker compose -f docker-compose.local.yml up -d tdp-frontend-test --quiet-pull
```

File: scripts/git-secrets-check.sh [hunk 7]
```diff
@@ -1,29 +1,55 @@
...
-        sudo cp /tmp/git-secrets/git-secrets /usr/sbin/
+        if [[ $(uname -s) == "Darwin" ]]; then  # Mac OSX check
+            gs_path="/usr/local/bin"
+        else # Linux, we're likely running in CircleCI
+            gs_path="/usr/sbin"
+        fi
+
+        echo "Moving git secrets into PATH"
+        sudo cp /tmp/git-secrets/git-secrets $gs_path/
+        $gs_path/git-secrets --install -f
```
Comment (andrew-jameson) [inline scripts/git-secrets-check.sh:20]: Add `rm -rf /tmp/git-secrets` and maybe change the availability check to `$gs_path` to make this more idempotent

File: Taskfile.yml (hunk 5 above)
Comment (andrew-jameson) [reply Taskfile.yml]: Unfortunately, these changes break my local use of Taskfile. We'll need to investigate your docker compose config so we're all on the same page. Additionally, just these 3 changes don't address what's needed for all other Tasks to run. Let's sync up when you're available

Comment (jtimpe) [reply Taskfile.yml]:
```bash
$ docker -v
Docker version 27.1.1, build 63125853e3
$ docker compose version
Docker Compose version v2.29.2-desktop.2
```

**Review by elipe17 — APPROVED** [review]

**Review by raftmsohani — APPROVED** [review]
LGTM

**Review by jtimpe — APPROVED** [review]

File: scripts/trufflehog-check.sh [hunk 8]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 set -e
 
-pythonVersion="3.10"
+pythonVersion="3"
```
Comment (andrew-jameson) [inline scripts/trufflehog-check.sh]:
```suggestion
pythonVersion="3.10"
```

File: scripts/trufflehog-check.sh [hunk 9]
```diff
@@ -25,7 +25,8 @@ last_merge=$(git log --format=format:"%H" -n 1 "$CIRCLE_BRANCH")
 
 # --since_commit - Look at all commits since the last merge into develop
 # --entropy=True - Entropy checks on large git diffs
-python ./trufflehog-check/lib/python$pythonVersion/site-packages/truffleHog/truffleHog.py \
+hogpath=$(find . -name truffleHog.py)
+python $hogpath \
```
Comment (andrew-jameson) [inline scripts/trufflehog-check.sh]:
```suggestion
python ./trufflehog-check/lib/python$pythonVersion/site-packages/truffleHog/truffleHog.py \
```

**Pull request merged**